### PR TITLE
[FFM-1363]: JS SDK handles segment updates incorrectly

### PR DESCRIPTION
**Pull request by davejohnston**
What:
When a patch event for a target-segment is received we fetch all flag variations
and generate a change event for flags whos variation has changed for this target.

Why:
When rules are changed the ff service will send an event to the client SDKs.
SDKs should then fetch the updated data. If flag rules change, then only that flag needs to
be fetched, but  If segment rules change, then all flag vairations must be fetched.

Previously when we received a patch event for a target segment, we tried to fetch an
invidual flag, and were passing the segment name instead of the flag name, so this resulted
in a 404.

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -279,9 +326,19 @@ const initialize = (apiKey: string, target: Target, options?: Options): Result =
 
       data.forEach((_evaluation: Evaluation) => {
         const _value = convertValue(_evaluation)
-        storage[_evaluation.flag] = _value
-        evaluations[_evaluation.flag] = { ..._evaluation, value: _value }
+
+        // Update the flag if the values are different
+        const _oldValue = storage[_evaluation.flag]
+        if (_value !== _oldValue) {
+          logDebug("Flag variation has changed for ", _evaluation.identifier)
```
Comment (knagurski) [inline src/index.ts:333]: Minor thing
```suggestion
          logDebug('Flag variation has changed for ', _evaluation.identifier)
```

File: src/index.ts [hunk 2]
```diff
@@ -165,6 +165,53 @@ const initialize = (apiKey: string, target: Target, options?: Options): Result =
...
+            }
+
+            return property === 'value' ? convertValue(evaluation) : evaluation[property]
+          }
+        })
+        : {
+          deleted: evaluation.deleted,
+          flag: evaluation.flag,
+          value: convertValue(evaluation)
+        }
```
Comment (knagurski) [inline src/index.ts:211]: Might just be personal preference, but this looks way too big to use a ternary IMHO. Consider using a more readable `if`/`else` instead

**Comment by knagurski** [comment]
@davejohnston with prerequisites a change in one flag could cascade to other flags. If this happens, are there events sent to the SDK to refetch all affected flags, or just the one that started the change?

**Review by chrisblakely01 — APPROVED** [review]
Looks good to me

File: src/index.ts (hunk 1 above)
Comment (davejohnston) [reply src/index.ts:333]: cool I can change that

File: src/index.ts (hunk 2 above)
Comment (davejohnston) [reply src/index.ts:211]: cool I'd refactored that out of an existing block and lifted as is, but make sense to change that for readability now

**Comment by davejohnston** [comment]
> @davejohnston with prerequisites a change in one flag could cascade to other flags. If this happens, are there events sent to the SDK to refetch all affected flags, or just the one that started the change?

Just checked the server code and sadly we don't handle pre-reqs , so its a single patch event.   Its outside the scope of this, but I'll raise a ticket as I don't think that was widely known, and we would need it to properly support pre-reqs.

**Pull request merged**